lib/oplog/last_minute_metric.go

```diff
@@ -0,0 +1,265 @@
+)
+
+// IntervalMaxMetric is a prometheus metric that reports the maximum value reported to it within a configurable
+// interval. These intervals are disjoint windows, and the *last* completed window is reported, if it immediately
+// precedes the current one.
+type IntervalMaxMetric struct {
+	desc        *prometheus.Desc
+	opts        IntervalMaxOpts
+	labelValues []string
+
+	lck sync.Mutex
+
+	currentMax  *lastMax
+	previousMax *lastMax
+}
+
+type lastMax struct {
+	value        float64
+	bucketedTime time.Time
+}
+
+// DefaultInterval is the default interval for the
+const DefaultInterval = 1 * time.Minute
+
+// IntervalMaxOpts are options for IntervalMaxMetric.
+type IntervalMaxOpts struct {
+	prometheus.Opts
+
+	// ReportInterval is the interval by which reports will be bucketed. Default 1m.
+	ReportInterval time.Duration
+}
+
+// NewIntervalMaxMetric constructs a new IntervalMaxMetric.
+func NewIntervalMaxMetric(opts IntervalMaxOpts, labels []string, labelValues []string) *IntervalMaxMetric {
+	if opts.ReportInterval == 0 {
+		opts.ReportInterval = DefaultInterval
+	}
+
+	return &IntervalMaxMetric{
+		desc: prometheus.NewDesc(
+			prometheus.BuildFQName(
+				opts.Namespace, opts.Subsystem, opts.Name,
+			),
+			opts.Help,
+			labels,
+			opts.ConstLabels,
+		),
+		opts:        opts,
+		labelValues: labelValues,
+
+		currentMax: nil,
+	}
+}
+
+// Describe implements the prometheus.Collector interface.
+func (c *IntervalMaxMetric) Describe(descs chan<- *prometheus.Desc) {
+	descs <- c.desc
+}
+
+// Collect implements the prometheus.Collector interface.
+func (c *IntervalMaxMetric) Collect(mtcs chan<- prometheus.Metric) {
+	c.lck.Lock()
+	defer c.lck.Unlock()
+
+	currentBucket := c.opts.thisTimeBucket()
+	c.rotate(currentBucket)
+
+	if c.previousMax == nil {
+		return
+	}
+
+	if currentBucket.Sub(c.previousMax.bucketedTime) != c.opts.ReportInterval {
+		return
+	}
+
+	mtcs <- prometheus.MustNewConstMetric(c.desc, prometheus.GaugeValue, c.previousMax.value, c.labelValues...)
+}
+
+// Report reports a value to the IntervalMaxMetric. If it is the greatest seen so far in the current window, it will be
+// recorded as such until either another, greater value is reported, or the window ends (it will then be the
+// value this IntervalMaxMetric returns when polled via Collect, until another window elapses).
+func (c *IntervalMaxMetric) Report(value float64) {
+	c.lck.Lock()
+	defer c.lck.Unlock()
+
+	thisTimeBucket := c.opts.thisTimeBucket()
+	c.rotate(thisTimeBucket)
+
+	maxVal := &lastMax{
+		value:        value,
+		bucketedTime: thisTimeBucket,
+	}
+
+	if c.currentMax == nil {
+		c.currentMax = maxVal
+		return
+	}
+
+	if thisTimeBucket.Equal(c.currentMax.bucketedTime) {
+		if c.currentMax.value < value {
+			c.currentMax = maxVal
+		}
+
+		return
+	}
+
+	if thisTimeBucket.After(c.currentMax.bucketedTime) {
+		c.currentMax = maxVal
+		return
+	}
+
+	// this bucket is before previous bucket
+	panic("interval max metric time traveled")
```

Comment:
I think we would be fine in the case of a leap second, since we'd never be reading time earlier than we expected, but yeah, an NTP sync would be a problem.